lib/Onnxifi/Base.cpp

```diff
@@ -99,141 +104,91 @@ void Event::wait() {
-                                         std::unique_ptr<glow::Context> ctx) {
-          afterRun(std::move(ctx));
-        });
-  } else {
-    executionEngine_.run(*ctx);
-    afterRun(std::move(ctx));
-  }
+Graph::~Graph() {
+  // Remove network from hostmanager
+  // backendPtr_->getHostManager().removeNetwork(netName_);
 }
 
-onnxStatus Graph::setIO(uint32_t inputsCount,
-                        const onnxTensorDescriptorV1 *inputDescriptors,
-                        uint32_t outputsCount,
-                        const onnxTensorDescriptorV1 *outputDescriptors) {
-  // Avoid race on setting inputs/outputs and graph run.
-  inputRunMutex_.lock();
+onnxStatus Graph::setIOAndRunAsync(
+    uint32_t inputsCount, const onnxTensorDescriptorV1 *inputDescriptors,
+    uint32_t outputsCount, const onnxTensorDescriptorV1 *outputDescriptors,
+    EventPtr outputEvent) {
 
-  // Build name to buffer mapping for inputs and outputs from scratch.
-  inputPlaceholderToBuffer_.clear();
-  outputPlaceholderToBuffer_.clear();
+  auto ctx = llvm::make_unique<Context>();
 
-  // Process inputs.
+  // Create tensors for input placeholders
   for (unsigned i = 0; i < inputsCount; ++i) {
-    const auto &in = inputDescriptors[i];
-    if (!onnxInputToPlaceholder_.count(in.name)) {
+    const auto &inOnnxTensor = inputDescriptors[i];
+    auto *inOnnxBuffer = reinterpret_cast<void *>(inOnnxTensor.buffer);
+
+    auto inPhIt = onnxInputToPlaceholder_.find(inOnnxTensor.name);
+    if (inPhIt == onnxInputToPlaceholder_.end()) {
       return ONNXIFI_STATUS_UNIDENTIFIED_NAME;
     }
 
-    auto *input = onnxInputToPlaceholder_[in.name];
-    inputPlaceholderToBuffer_.insert({input, in.buffer});
+    auto &inPhPtr = inPhIt->getValue();
+
+    auto t = llvm::make_unique<Tensor>(inOnnxBuffer, inPhPtr->getType());
+
+    ctx->insert(inPhPtr, std::move(t));
   }
 
-  // Process outputs.
+  // Create tensors for output placeholders
   for (unsigned i = 0; i < outputsCount; ++i) {
-    const auto &out = outputDescriptors[i];
+    const auto &outOnnxTensor = outputDescriptors[i];
+    auto *outOnnxBuffer = reinterpret_cast<void *>(outOnnxTensor.buffer);
 
-    if (!onnxOutputToPlaceholder_.count(out.name)) {
+    auto outPhIt = onnxOutputToPlaceholder_.find(outOnnxTensor.name);
+    if (outPhIt == onnxOutputToPlaceholder_.end()) {
       return ONNXIFI_STATUS_UNIDENTIFIED_NAME;
     }
 
-    auto *output = onnxOutputToPlaceholder_[out.name];
-    outputPlaceholderToBuffer_.insert({output, out.buffer});
+    auto &outPhPtr = outPhIt->getValue();
+
+    auto t = llvm::make_unique<Tensor>(outOnnxBuffer, outPhPtr->getType());
+
+    ctx->insert(outPhPtr, std::move(t));
   }
 
+  // Run
+  getHostManager().runNetwork(
+      netName_, std::move(ctx),
+      [outputEvent](runtime::RunIdentifierTy runId, runtime::ResultCode result,
+                    std::unique_ptr<Context> ctx) { outputEvent->signal(); });
```

Comment:
Are you doing anything with the returned context? Otherwise I think it will get freed.

Reply:
I don't think there's anything that needs to be done with the ctx here right? I think it should be just freed.